### ISSUE TITLE
Estimating reference expression signatures from very large dataset

### ISSUE THREAD

**Issue by AlexanderAivazidis** [BayraktarLab/cell2location#33]
Hi,

I am trying to estimate reference expression signatures from this very large dataset from the Linnarson lab:

http://mousebrain.org/downloads.html

I am loading there l5_all.loom file as an AnnData object:

loom = sc.read_loom('../data/Zeissel/l5_all.loom')

Then I am selecting some relevant brain regions and go through your recommended preprocessing steps:

loom.obs['sample_name_col'] = loom.obs_names
loom.obs['CombinedClusterName'] = [loom.obs['Class'].iloc[i] + '_' + str(loom.obs['Clusters'].iloc[i]) + '_' + loom.obs['ClusterName'].iloc[i] for i in range(len(loom.obs['ClusterName']))] 
loom.var['SYMBOL'] = loom.var.index
irrelevant_regions = np.array(('Dorsal root ganglion', 'Dorsal root ganglion,Sympathetic ganglion', 'Enteric nervous system',
                               'Medulla', 'Spinal cord', 'Sympathetic ganglion'))
subset_regions =[r not in irrelevant_regions for r in loom.obs['Region']]
loom = loom[subset_regions,:]
print(np.shape(loom))
import matplotlib as mpl
# remove cells and genes with 0 counts everywhere
sc.pp.filter_cells(loom, min_genes=1)
sc.pp.filter_genes(loom, min_cells=1)

# calculate the mean of each gene across non-zero cells
loom.var['n_cells'] = (loom.X.toarray() > 0).sum(0)
loom.var['nonz_mean'] = loom.X.toarray().sum(0) / loom.var['n_cells']

plt.hist2d(np.log10(loom.var['nonz_mean']),
           np.log10(loom.var['n_cells']), bins=100,
           norm=mpl.colors.LogNorm(),
           range=[[0,0.5], [1,4.5]]);

nonz_mean_cutoff = np.log10(1.12) # cut off for expression in non-zero cells
cell_count_cutoff = np.log10(loom.shape[0] * 0.0005) # cut off percentage for cells with higher expression
cell_count_cutoff2 = np.log10(loom.shape[0] * 0.03)# cut off percentage for cells with small expression

plt.vlines(nonz_mean_cutoff, cell_count_cutoff, cell_count_cutoff2, color = 'orange');
plt.hlines(cell_count_cutoff, nonz_mean_cutoff, 1, color = 'orange');
plt.hlines(cell_count_cutoff2, 0, nonz_mean_cutoff, color = 'orange');
plt.xlabel('Mean count in cells with mRNA count > 0 (log10)');
plt.ylabel('Count of cells with mRNA count > 0 (log10)');
loom[:,(np.array(np.log10(loom.var['nonz_mean']) > nonz_mean_cutoff)
         | np.array(np.log10(loom.var['n_cells']) > cell_count_cutoff2))
      & np.array(np.log10(loom.var['n_cells']) > cell_count_cutoff)].shape
# select genes based on mean expression in non-zero cells
loom = loom[:,(np.array(np.log10(loom.var['nonz_mean']) > nonz_mean_cutoff)
         | np.array(np.log10(loom.var['n_cells']) > cell_count_cutoff2))
      & np.array(np.log10(loom.var['n_cells']) > cell_count_cutoff)
              & np.array(~loom.var['SYMBOL'].isna())]

The size of the final dataset is (142160, 11812).
Now when I run your regression it just gets stuck:

r, adata_snrna_raw = run_regression(loom, # input data object]

                   verbose=True, return_all=True,

                   train_args={
                    'covariate_col_names': ['CombinedClusterName'], # column listing cell type annotation
                    'sample_name_col': 'sample_name_col', # column listing sample ID for each cell
                    # column listing technology, e.g. 3' vs 5',
                    # when integrating multiple single cell technologies corresponding
                    # model is automatically selected
                    'tech_name_col': None,

                    'stratify_cv': 'CombinedClusterName', # stratify cross-validation by cell type annotation

                    'n_epochs': 100, 'minibatch_size': 1024, 'learning_rate': 0.01,

                    'use_cuda': True, # use GPU?

                    'train_proportion': 0.9, # proportion of cells in the training set (for cross-validation)
                    'l2_weight': True,  # uses defaults for the model

                    'readable_var_name_col': 'SYMBOL', 'use_raw': False},

                   model_kwargs={}, # keep defaults
                   posterior_args={}, # keep defaults

                   export_args={'path': results_folder + 'regression_model/', # where to save results
                                'save_model': False, # save pytorch model?
                                'run_name_suffix': ''})

reg_mod = r['mod']

In particular it just shows:

Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
Variable names are not unique. To make them unique, call `.var_names_make_unique`.
### Creating model ### - time 0.11 min

And nothing else. If I subset the loom object to 1000genes x 1000cells everything runs fine and I see the optimization progress bar straight away. I have increased the RAM to 500GB, but still it just gets stuck and does not show the optimization progress. This is now still the same after 12+ hours after starting. I am using the GPU so that should not be the problem.

So I wonder if you have any suggestions for this. Otherwise, splitting the data up into smaller chunks would be an option.

Thanks!

Alexander

**Comment by vitkl** [comment]
Hi @AlexanderAivazidis 

There are a few potential issues:

1. You need unique gene names and unique cell names. That could be the issue with the training being stuck.

2. The problem is that with 142,160 cells, `'train_proportion': 0.9` leads to a very large validation set (14,216) which can lead to running out of GPU memory or slow training. You need to reduce validation set proportion, e.g. `'train_proportion': 0.97`. In the future, I am planning to move this step to scVI which deal with this better.
